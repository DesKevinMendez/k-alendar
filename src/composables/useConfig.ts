import { Settings } from "luxon";
import { readonly, ref } from "vue";
const langCalendar = ref('en')
const lang = readonly(langCalendar)

export default function useConfig() {

  const setLang = (langStr: string) => {
    langCalendar.value = langStr;
    Settings.defaultLocale = langStr;
  }

  return {
    lang,
    setLang
  }
}