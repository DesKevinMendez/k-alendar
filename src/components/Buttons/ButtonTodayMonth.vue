<template>
  <button type="button" class="k-alendar-today-button" @click="handleClick">{{ text }}</button>
</template>

<script setup lang="ts">
import useConfig from '@/composables/useConfig'
import useDate from '@/composables/useDate'
import useRenderCalendar from '@/composables/useRenderCalendar'
import { translations } from '@/locales/traductions';
import { DateTime } from 'luxon'
import { computed } from 'vue';
const emit = defineEmits(['handle'])
const { currentDate, monthDays, generateCalendar } = useRenderCalendar()
const { today } = useDate()
const { lang } = useConfig()

const handleClick = () => {
  currentDate.value = DateTime.now()

  monthDays.value = generateCalendar(today.value)

  emit('handle', today.value)
}

const text = computed(() => {
  return translations[lang.value]?.buttons.today || 'Today'
})
</script>

<style scoped lang="postcss">
.k-alendar-today-button {
  @apply px-2 py-1 border border-gray-200 rounded-md
  lg:p-2
  dark:border-slate-600;
}
</style>
