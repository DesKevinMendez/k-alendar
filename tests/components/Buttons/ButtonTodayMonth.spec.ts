
import ButtonTodayMonth from '@/components/Buttons/ButtonTodayMonth.vue';
import useConfig from '@/composables/useConfig';
import { mount, VueWrapper } from '@vue/test-utils';
import { DateTime, Settings } from 'luxon';
import { Date2024Oct23 } from '../../fakes/monthDays';
const { setLang } = useConfig()

const dateToTest = '2024-09-10';
const mockGenerateCalendar = vi.fn().mockReturnValue(Date2024Oct23);
const currentDate = { value: DateTime.fromISO(dateToTest) };
const monthDays = { value: [] };

vi.mock('@/composables/useRenderCalendar', () => ({
  default: vi.fn(() => ({
    currentDate,
    monthDays,
    generateCalendar: mockGenerateCalendar,
  })),
}));

let wrapper: VueWrapper<InstanceType<typeof ButtonTodayMonth>>

describe('ButtonTodayMonth', () => {

  beforeEach(() => {
    wrapper = mount(ButtonTodayMonth);
  });

  afterEach(() => {
    vi.clearAllMocks()
  })

  it('should render the component', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('init values for composable useRenderCalendar', () => {
    expect(monthDays.value).toStrictEqual([]);
    expect(currentDate.value).toStrictEqual(DateTime.fromISO(dateToTest));
  })

  it('do click on the button', async () => {
    const today = '2024-09-23'

    // mock DateTime.now()
    const t = DateTime.fromISO(today);
    Settings.now = () => t.toMillis();

    const button = wrapper.find('button');
    await button.trigger('click');

    expect(monthDays.value).toStrictEqual(Date2024Oct23);
    expect(currentDate.value).toStrictEqual(DateTime.fromISO(today));
    expect(mockGenerateCalendar).toHaveBeenCalledWith(today);
    expect(wrapper.emitted('handle')).toStrictEqual([[today]]);

    await wrapper.find('button').trigger('click');
    expect(wrapper.emitted('handle')).toStrictEqual([[today], ['2024-09-23']]);
    expect(currentDate.value).toStrictEqual(DateTime.fromISO('2024-09-23'));
  })

  describe('computed text', () => {
    it('by default', async () => {
      const button = wrapper.find('button');

      expect(button.text()).toBe('Today');
    })

    it('is es', async () => {
      setLang('es');
      await wrapper.vm.$nextTick();

      const button = wrapper.find('button');

      expect(button.text()).toBe('Hoy');
    })

    it('is it', async () => {
      setLang('it');
      await wrapper.vm.$nextTick();

      const button = wrapper.find('button');

      expect(button.text()).toBe('Oggi');
    })

    it('lang is invalid, it should be english', async () => {
      setLang('invalid');
      await wrapper.vm.$nextTick();

      const button = wrapper.find('button');

      expect(button.text()).toBe('Today');
    })
  })
});