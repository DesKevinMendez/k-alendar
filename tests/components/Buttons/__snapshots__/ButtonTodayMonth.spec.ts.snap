// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ButtonTodayMonth > should render the component 1`] = `"<button data-v-8f6b2986="" type="button" class="k-alendar-today-button">Today</button>"`;
